# DETAILED DESCRIPTION OF EACH MODULE

## Read_qc module

The MuDoGeR Read_qc module is based on MetaWrap Read_qc module and is responsible to prepare the raw reads for assembly and alignment. The trimming of the raw reads is based on adapted content and PHRED scored with the default setting of Trim-galore. As a result, all the non-high quality sequences are removed. The high quality reads that remained are aligned to the genome of the host with bmtagger leading to removal of host contamination and the read pairs with only one aligned read from the metagenomic data. In the final step, , FASTQC is used to generate quality reports of the raw and final read sets in order to assess read quality improvement. The users have control over which of the above features they wish to use.

## Calculation of resources module
Before the user goes to the assembly module, it is possible to estimate the necessary resources for the assembly of the library after produced in quality control module. This can be achieved by estimating first the complexity of the forward orreverese pure reads by k-mer calculation, as they are almost identical. The k-mer is usually estimated in size 33 or 55. Both values should be estimated. Then k-mer with size 33 and 55 are combined to a single  output file. Using this file the necessary amount of memory is predicted. Finally, the memory usage (in megabytes) is fetced. 


## Assembly module
The  MuDoGeR Assembly module is based on MetaWrap Assembly module. This module is utillized for assembling a set of metagenomic reads. The assembly can be achieved by using either MegaHit or metaSPAdes tools. Although, MegaHit is able to assemble large and complex metagenomic data and is cost memory and time-efficient metaSPAdes assembly is usually superior. The MegaHit assembly is sorted by length and their names are changed to resembled the names in the SPAdes, including the contig ID, length, and coverage. In the end the scafgolds with lenght <1000bp are removed and QUAST formates an assembly report.

## Prokaryotic module
The MuDoGeR prokaryotic module is a combination of modules included in the MetaWrap tool. The module starts with the binning of the assembly datasets with the use of three wrapped tools (**MaxBin2**, **metaBAT2**, **CONCOCT**). The process results to the production of bin files from each method. The next step is the bin refinement of the bins produced from each tool. As a result, bin sets creating. In the same step the **CheckM** tool is running for the evaluation of the qualitiy of the refined  bin set in terms of completion/contamination, with 30/40 for Archaea genomes and 50/10 for Bacterial genomes. After this, there is an optional step, where the user can choose to  Next, the produced bins from Archaea and Bacteria are re-evaluated with the respective completion/contamination score creating a final Archaea bin set and a final Bacteria bin set. For the classification of the bin sets from Bacteria and Archaea produced in the bin refinement module, **GTDB** tool is used. Next, the prediction of the functional annotation of the re-assembled bin sets,is achieved by **PROKKA**, which utilizes a number of softawre tools. Full detailed description of the Prokaryotic module can be found in ![MetaWrap/Module_descriptions.md](https://github.com/bxlab/metaWRAP/blob/master/Module_descriptions.md). Then, a selection of prokaryotic MAGs takes place. Finally, the **uBin** refining tool is used to for the manual curation of genomes from the selected metagenomic bins 

## Viral module
The MuDoGeR viral module utilizes three viral recovery tools(**VirFinder**, **VirFinder**, **VIBRANT**) for the identification and recovery of viral genomes from an assemble metagenomic dataset. In the beginning, each of the tools is used to recover independently the viral genomes from a given assembly dataset and locates them in seperate folders.  In the next step the recovered sequences are filtered using bash. The selection of proper VirFinder sequences is based on low q-value (q-value=< 0.01) and high length (length>= 1000 bp), the VirSorter chosen sequences are those classified into categories 1 and 2 and from VIBRANT the selected sequences are contained in the Assembly.phages_combined.fna file. The headers of VirSorter and VIBRANT contigs are modified so they can match with those produced from the VirFinder recovery.It is important to note that in contrast with the other two tools, the VirFinder output recovery file contains only the headers of the assemblies. Because of that,using bash, the headers of the unique filtered sequences from each tool are extracted to a common fasta.file and sorted by length. In the next step, using the headers including in the common fasta.file the actual sequences from the assembly dataset are extracted transfered to a new fasta.file with python. Next,the duplicated contigs are removed by a de-replication function. The dependencies of the dereplication step are  maximum coverage (-c 70) and maximum identity (-i 95). The dereplication is followed by quality estimation and taxonomy of the the dereplicated contigs by **CheckV** and **vContact2** respectively. Then, the host identification of dereplicated contigs is taking place by using **WIsH tool**. Finally, a selection of viral  MAGs takes place.

## Eukaryotic module
In the first step on the Eukaryotes recovery module, the assembled data-set is separated to Prokaryotic and Eukaryotic using **EukRep** tool. In case the size of the Eukaryotic assembly file is >= 2.5 Mb, the Eukaryotic recovery modules can continue to the second step, the automated binnning with **CONCOCT** tool. The eukaryotic bins  produced by **CONCOCT** are then filtered. Bins with size <= 2.5 Mb are removed. After the filtering a chain of processes follows: 1)**GeneMark-ES** is responsible for the gene prediction of the remaining bins. 2)The predicted genes from **GeneMark-ES** are annotated with **Maker2** and 3) their quality of the annotated proteins is measured by estimating the completeness using  **BUSCO**. Finally, the contamination of the bins which were kept after the filtering is estimated by using **EukCC tool**. Finally, a selection of eukaryotic MAGs takes place.

