# DETAILED DESCRIPTION OF EACH MODULE

## Read_qc module

The Read_qc module is responsible to preparethe raw reads for assembly and alignment. The trimming of the raw reads is based on adapted content and PHRED scored with the default setting of Trim-galore. As a result, all the non-high quality sequences are removed. The high quality reads that remained are aligned to the genome of the host with bmtagger leading to removal of host contamination and the read pairs with only one aligned read from the metagenomic data. In the final step, , FASTQC is used to generate quality reports of the raw and final read sets in order to assess read quality improvement. The users have control over which of the above features they wish to use.




## Viral module
The viral module utilizes three viral recovery tools(**VirFinder**, **VirFinder**, **VIBRANT**) for the identification and recovery of viral genomes from an assemble metagenomic dataset. In the beginning, each of the tools is used to recover independently the viral genomes from a given assembly dataset and locates them in seperate folders.  In the next step the recovered sequences are filtered using bash. The selection of proper VirFinder sequences is based on low q-value (q-value=< 0.01) and high lentgh (lentgh>= 1000 bp), the VirFinder chosen sequences are those classified into categories 1 and 2 and from VIBRANT the selected sequences are contained in the Assembly.phages_combined.fna file. It is important to note that in contrast with the other two tools, the VirFinder output recovery file contains only the headers of the assemblies. Because of that,with bash command, the headers of the uniq filtered sequences from each tool are extracted to a common fasta.file. In the next step, using the headers including in the common fasta.file,  the actual sequences from the assembly dataset are extracted transfered to a new fasta.file with python. Finally,the duplicated contigs are removed by a de-replication function. The dependencies of the dereplication step are  maximum coverage (-c 70) and maximum identity (-i 95). The values of the dereplication parameters can be changed depending on the users aim and datasets. The final output results, are depositet to a new output folder, contained inside the initial output directory.
